### PR TITLE
RSDK-5244 Make the AMS encoder talk to i2c bus directly

### PR DESCRIPTION
**Pull request by oliviamiller**
tested on scuttle AMS encoders. This change also needs a change in app to remove board from the "fancy config" and a migration script.

### CODE REVIEW

**Review by penguinland — COMMENTED** [review]
LGTM, though we should make sure not to merge this until a config migration is ready. Alas, this component's config has a different structure than the others I've done, so the migration won't be just a 1-line change like https://github.com/viamrobotics/app/blob/main/etc/migrations/raw_i2c_spi_access/raw_i2c_spi_access.go#L98. but hopefully a lot of that code can still be useful.

File: components/encoder/ams/ams_as5048.go
Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:142]: (or, tell me what your maximum line width is and that merging the lines would violate it)

Comment (oliviamiller) [reply components/encoder/ams/ams_as5048.go:142]: will change to one arg per line

File: components/encoder/ams/ams_as5048.go
Comment (oliviamiller) [reply components/encoder/ams/ams_as5048.go]: I like that implementation better! updated

File: components/encoder/ams/ams_as5048.go
Comment (oliviamiller) [inline components/encoder/ams/ams_as5048.go:180]: @penguinland  Just a note that I noticed when testing- the `genericlinux.NewI2cBus` call does **not** error if the i2c bus does not exist, so this error message would never be hit. Is this expected? In this implementation, if I put an invalid i2cBus, it successfully reconfigures and the code errors when calling `openHandle` in `resetPositon`. In the old implementation an invalid i2c bus fails reconfiguration at the `localBoard.I2CByName` call. Is this an issue?

Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:180]: Hm... good question. The distinction is that the old way relied on the board config naming the I2C bus, and would fail if the name didn't exist. but if you named a nonexistent bus, I think this part would not have errors and then you'd get errors later when you actually try using the bus. I think that's analogous to the current behavior. 

It could be convenient to have `NewI2cBus` run some kind of check and return errors if the bus doesn't exist, but off the top of my head I don't see what that check would look like: we don't know which address to talk to (perhaps multiple addresses?) nor what we'd say to them. We could try creating something reminiscent of `i2cdetect`, but that sounds like a giant pain.

So, I'm inclined to say that this isn't an issue, but I wish we could more reliably return an error here. :confused:

File: components/encoder/ams/ams_as5048.go
Comment (penguinland) [inline components/encoder/ams/ams_as5048.go:177]: I grant that these are the conditions in which we _need_ to initialize a new bus, but is it ever harmful to do so? Like, what if we removed the if statement and re-initialized it every time? The code would be less complicated, and we wouldn't need to store the bus name any more.

I haven't thought this through: maybe there's a good reason to keep the old bus around if it hasn't changed. Keep the PR as-is if you want.

**Review by penguinland — COMMENTED** [review]
LGTM, though we still need a config migration in the app repo before this gets merged.

File: components/encoder/ams/ams_as5048.go
Comment (randhid) [inline components/encoder/ams/ams_as5048.go:142]: can we clean up the constructor so it is of the form:
```Constructor: newAmsEncoder```
rather than
```Constructor: func(.....){...}```

File: components/encoder/ams/ams_as5048.go
Comment (randhid) [inline components/encoder/ams/ams_as5048.go:126]: can we clean up the constructor so it is of the form:
```Constructor: newAmsEncoder```
rather than
```Constructor: func(.....){...}```

Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:126]: I agree with the sentiment of your comment, but don't see where it goes. The only mention of `Constructor` in this file already has that form. Did you perhaps mean to put this comment on a different PR?

Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:126]: Rand and I discussed: this comment was left by mistake.

File: components/encoder/ams/ams_as5048.go
Comment (randhid) [reply components/encoder/ams/ams_as5048.go:180]: let's check with the ina219 sensor which relies completely on an external library to open its i2c bus

Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:180]: The ina219 uses [this function](https://github.com/d2r2/go-i2c/blob/master/i2c.go#L31-L41), which immediately tries opening the device (file in `/dev`) and telling it what address we're going to connect to. It returns an error if the device cannot be opened or written to. 

Part of me thinks we should change `genericlinux.NewI2cBus` to use the same library the ina219 uses: it would be more consistent to use just one I2C library throughout the RDK. Part of me thinks that updating that is nontrivial (specifically, the ina's library expects the bus ID to be an int, while genericlinux/Periph's version wants a string that can include a `"/dev/i2c-"` prefix), and the consistency is not worth a breaking change.

If we do want to change that up, I think it's a separate enough issue to warrant its own ticket/PR, and the current work should not be blocked by it.

Comment (penguinland) [reply components/encoder/ams/ams_as5048.go:180]: As a stopgap measure to split the difference, we could add into `genericlinux.NewI2cBus` a thing to get the longest numeric suffix of the string passed in, prepend a `/dev/i2c-` to the front, and try opening that file, just to see if we get permission errors. It won't work if you're passing in a non-numeric string (e.g., you've set up udev to create semantic symlinks like `/dev/i2c-humidity` and put that in the config), but I doubt many people do that (and we can check in the config database to see how common it really is). but again, I think changing that should be a separate PR

Comment (randhid) [reply components/encoder/ams/ams_as5048.go:180]: Agree that changing that is a separate pr and separate ticket altogether, but erroring out at the i2c/spi handle opening step is important.

**Review by randhid — APPROVED** [review]
okay to merge - I know olivia tested it out with the scuttle before any changes to this pr, maybe worth a second look with hardware.

**Comment by penguinland** [comment]
Waiting on https://github.com/viamrobotics/app/pull/2958: we don't want to merge this until the config migration for it is ready.

**Comment by penguinland** [comment]
The migration is merged, and the hardware still works!

**Pull request merged**
